operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java

```diff
@@ -564,47 +563,143 @@ private void setStatusConditions(DomainStatus status) {
         }
       }
 
-      private void addTooManyReplicasFailures(DomainStatus status) {
-        getConfiguredClusters().stream()
-              .map(TooManyReplicasCheck::new)
-              .filter(TooManyReplicasCheck::isFailure)
-              .forEach(check -> status.addCondition(check.createFailureCondition()));
+      private boolean haveServerData() {
+        return this.serverState != null;
       }
 
-      private List<WlsClusterConfig> getConfiguredClusters() {
-        return Optional.ofNullable(config).map(WlsDomainConfig::getConfiguredClusters).orElse(Collections.emptyList());
+      class Conditions {
+
+        private final ClusterCheck[] clusterChecks = createClusterChecks();
+        private final List<DomainCondition> conditions = new ArrayList<>();
+
+        void apply(DomainStatus status) {
+          status.removeConditionsMatching(c -> c.hasType(Failed) && ReplicasTooHigh.name().equals(c.getReason()));
+          conditions.forEach(status::addCondition);
+        }
+
+        List<DomainCondition> getNewConditions(DomainStatus status) {
+          return conditions.stream()
+                .filter(c -> "True".equals(c.getStatus()))
+                .filter(c -> status == null || !status.hasConditionWith(matchFor(c)))
+                .collect(Collectors.toList());
+        }
+
+        Predicate<DomainCondition> matchFor(DomainCondition condition) {
```

Comment:
Shall we take the failure reason into consideration as well in the logic here when the condition type is Failed?

Reply:
Can you explain? Take it into consideration how?

Reply:
When matchFor method checks if one condition matches another condition, currently it only take the condition type into consideration. When the type is "Failed",  we have many different failure reasons, such as DomainInvalid, Kubernetes, Internal, ReplicasTooHigh and etc. Should a "Failed" condition with one failure reason be considered different from another "Failed" condition with a different failure reason?

Reply:
The purpose of this method is to return the new conditions for which we might generate events. Currently, that means only Available and Completed, as we have other places that generate Failure events. If we want to change that so this is also the source of the Failure events, we will indeed need to make that change.